src/gnatss/ops/solve.py

```diff
@@ -0,0 +1,309 @@
+
+    # Add travel times variance on diagonal only
+    n = len(covariance_matrix)
+    for i in range(n):
+        covariance_matrix[i, i] = covariance_matrix[i, i] + travel_times_variance
+    return covariance_matrix
+
+
+@numba.njit(cache=True)
+def _calc_weight_fac(transponders_mean_sv):
+    """
+    Calculate the weight factor array
+
+    .. math::
+        W_f = \\frac{2.0}{s^2}
+    """
+    return 2.0 / (transponders_mean_sv**2)
+
+
+@numba.njit(cache=True)
+def _calc_weight_mat(covariance_std):
+    """
+    Calculate the weight matrix by inverting the covariance std matrix
+    """
+    return np.linalg.inv(covariance_std)
+
+
+@numba.njit(cache=True)
+def __get_diagonal(array: NDArray[Shape["*, *"], Float64]) -> NDArray:
+    """
+    Get the diagonal of an array
+
+    Parameters
+    ----------
+    array : (N,M) ndarray
+        The input array to get diagonal from,
+        array shape must be even
+
+    Returns
+    -------
+    (N,) ndarray
+        The diagonal values of the input array
+    """
+    len_cm = len(array)
+    diag = np.zeros(len_cm)
+    for i in range(len_cm):
+        diag[i] = array[i, i]
+    return diag
+
+
+@numba.njit(cache=True)
+def calc_twtt_model(
+    transmit_vectors: NDArray[Shape["*, 3"], Float64],
+    reply_vectors: NDArray[Shape["*, 3"], Float64],
+    transponders_mean_sv: NDArray[Shape["*"], Float64],
+) -> NDArray[Shape["*"], Float64]:
+    """
+    Calculate the Modeled TWTT (Two way travel time) in seconds
+
+    .. math::
+        \\frac{\\hat{D_s} + \\hat{D_r}}{c}
+
+    Parameters
+    ----------
+    transmit_vector : (N,3) ndarray
+        The transmit array of vectors
+    reply_vector : (N,3) ndarray
+        The reply array of vectors
+    transponders_mean_sv : (N,) ndarray
+        The transponders mean sound speed
+
+    Returns
+    -------
+    (N,) ndarray
+        The modeled two way travel times in seconds
+
+    """
+    # Calculate distances in meters
+    transmit_distance = np.array(
+        [np.linalg.norm(vector) for vector in transmit_vectors]
+    )
+    reply_distance = np.array([np.linalg.norm(vector) for vector in reply_vectors])
+
+    return (transmit_distance + reply_distance) / transponders_mean_sv
+
+
+@numba.njit(cache=True)
+def calc_tt_residual(
+    delays, transponder_delays, twtt_model
+) -> NDArray[Shape["*"], Float64]:
+    """
+    Calculate the travel time residual in seconds
+
+    Parameters
+    ----------
+    delays : (N,) ndarray
+        The measured travel time delays from data (sec)
+    transponder_delays : (N,) ndarray
+        The set transponder delays as defined in configuration file (sec)
+    twtt_model : (N,) ndarray
+        The modeled two way travel times (sec)
+
+    Returns
+    -------
+    (N,) ndarray
+        The travel time residual
+    """
+
+    # dA = Ameas - Amod
+    return (delays - transponder_delays) - twtt_model
+
+
+@numba.njit(cache=True)
+def solve_transponder_locations(
+    transmit_xyz: NDArray,
```

Comment:
In other places you are adding the NDArray shape as part of the type annotation. Any reason these don't have it?

Reply:
No reason, I haven't gotten to this function yet for the specifity 😬